source/common/formatter/substitution_format_string.cc

```diff
@@ -18,15 +19,18 @@ SubstitutionFormatStringUtils::createJsonFormatter(const ProtobufWkt::Struct& st
 }
 
 FormatterPtr SubstitutionFormatStringUtils::fromProtoConfig(
-    const envoy::config::core::v3::SubstitutionFormatString& config, Api::Api& api) {
+    const envoy::config::core::v3::SubstitutionFormatString& config,
+    Server::Configuration::CommonFactoryContext& context) {
   // Instantiate formatter extensions.
   std::vector<CommandParserPtr> commands;
   for (const auto& formatter : config.formatters()) {
     auto* factory = Envoy::Config::Utility::getFactory<CommandParserFactory>(formatter);
     if (!factory) {
       throw EnvoyException(absl::StrCat("Formatter not found: ", formatter.name()));
     }
-    auto parser = factory->createCommandParserFromProto(formatter.typed_config());
+    auto typed_config = Envoy::Config::Utility::translateAnyToFactoryConfig(
+        formatter.typed_config(), context.messageValidationVisitor(), *factory);
```

Comment:
I am OK with this, we did however have a recent discussion about this in #14855 with @alyssawilk and @mattklein123 and there are two schools of thought as to where the translation should happen:

a) centrally, where the extension point instantiates the extensions or 
b) within the extension's config code so validations can be nicely unit tested without requiring the extension point code itself

Reply:
Ah ok, I see. 

For me it took quite a long time to figure out what was going on, since there isn't other extension points doing the translation in the extension itself, but that would of course get easier when there are examples.

`context` is not passed to formatter extension though, so it cannot get access to the message validation visitor?

Reply:
Another aspect: if extension are passed `typed_config` constructed as `Any` and extensions are expected to translate that to concrete config type, then does it still make sense to have extensions implement `TypedFactory::createEmptyConfigProto()`? Isn't the purpose of that method and the reason for inheriting from `TypedFactory` targeted to allow translation to happen centrally but still in a generic way?

Reply:
I think I prefer this way -- it's easier to write extensions. As long as there's a test that shows that validation is triggered in this path, we're probably fine. Though the other requirement is that something has to include the validation implementation for each config type so that the validation is actually triggered.